File: assets/js/reports.js
// ============================================================
// REPORTS PAGE
// ============================================================
var _reportsData = null;
var _admChart = null;

async function initReports() {
  var sel = document.getElementById('reportYear');
  if (!sel) return;
  if (!sel.options.length) {
    var yr = new Date().getFullYear();
    for (var y = yr; y >= yr - 3; y--) {
      sel.innerHTML += '<option value="' + y + '">' + y + '</option>';
    }
  }
  await loadReports();
}

async function loadReports() {
  var sel = document.getElementById('reportYear');
  var year = (sel ? sel.value : '') || new Date().getFullYear();
  var errEl = document.getElementById('reportsError');
  if (errEl) errEl.style.display = 'none';
  try {
    _reportsData = await api('/reports?year=' + year);
  } catch(e) {
    if (errEl) { errEl.textContent = 'Failed to load reports: ' + e.message; errEl.style.display = 'block'; }
    console.error(e);
    return;
  }
  // Update stats first (separate try so chart error doesn't hide them)
  try { renderReportStats(_reportsData); } catch(e) { console.error('stat error', e); }
  try { renderReportChart(_reportsData); } catch(e) { console.error('chart error', e); }
  try { renderReportTables(_reportsData); } catch(e) { console.error('table error', e); }
}

function renderReports(d) {
  renderReportStats(d);
  try { renderReportChart(d); } catch(e) { console.error('chart error', e); }
  renderReportTables(d);
}

function renderReportStats(d) {
  var rThis = document.getElementById('rThisMonth');
  var rLast = document.getElementById('rLastMonth');
  var rTot  = document.getElementById('rTotal');
  var rPend = document.getElementById('rPending');
  var rYear = document.getElementById('rChartYear');
  if (rThis) rThis.textContent = '₹' + (d.thisMonth||0).toLocaleString('en-IN');
  if (rLast) rLast.textContent = '₹' + (d.lastMonth||0).toLocaleString('en-IN');
  if (rTot)  rTot.textContent  = '₹' + (d.totalCollected||0).toLocaleString('en-IN');
  if (rPend) rPend.textContent = '₹' + (d.totalDue||0).toLocaleString('en-IN');
  if (rYear) rYear.textContent = d.year;
}

function renderReportChart(d) {
  // Monthly admissions chart
  var months = ['Jan','Feb','Mar','Apr','May','Jun','Jul','Aug','Sep','Oct','Nov','Dec'];
  var counts  = (d.monthlyAdmissions || []).map(function(m){ return m.count; });
  var chartEl = document.getElementById('admissionsChart');
  if (!chartEl) return;
  if (typeof Chart === 'undefined') {
    chartEl.parentNode.innerHTML = '<div style="text-align:center;padding:40px;color:var(--muted);">Chart.js not loaded — check your internet connection.</div>';
    return;
  }
  var ctx = chartEl.getContext('2d');
  if (_admChart) _admChart.destroy();
  _admChart = new Chart(ctx, {
    type: 'bar',
    data: {
      labels: months,
      datasets: [{
        label: 'Students Enrolled',
        data: counts,
        backgroundColor: 'rgba(59,130,246,0.7)',
        borderRadius: 6,
      }]
    },
    options: {
      responsive: true,
      plugins: { legend: { display: false } },
      scales: { y: { beginAtZero: true, ticks: { stepSize: 1 } } }
    }
  });
}

function renderReportTables(d) {

  // Course-wise table
  var cwEl = document.getElementById('courseWiseTable');
  if (cwEl) {
    cwEl.innerHTML = (d.courseWise || []).map(function(c) {
      return '<div style="display:flex;justify-content:space-between;align-items:center;padding:10px 0;border-bottom:1px solid var(--border);">' +
        '<div><span class="badge badge-primary">' + c.course + '</span><span style="margin-left:8px;font-size:.82rem;color:var(--muted);">' + c.studentCount + ' students</span></div>' +
        '<div style="font-weight:700;color:var(--success);">₹' + (c.collected||0).toLocaleString('en-IN') + '</div>' +
      '</div>';
    }).join('') || '<div style="color:var(--muted);text-align:center;padding:20px;">No data</div>';
  }

  // Serious defaulters
  var sdEl = document.getElementById('seriousDefaultersTable');
  if (sdEl) {
    sdEl.innerHTML = (d.seriousDefaulters || []).map(function(s) {
      return '<div style="padding:10px 0;border-bottom:1px solid var(--border);">' +
        '<div style="display:flex;justify-content:space-between;">' +
          '<div style="font-weight:600;">' + s.name + '</div>' +
          '<div style="color:var(--danger);font-weight:700;">₹' + s.due.toLocaleString('en-IN') + '</div>' +
        '</div>' +
        '<div style="font-size:.78rem;color:var(--muted);">' + s.course + ' · ' + s.monthsOverdue + ' months overdue · ' + (s.contact||'–') + '</div>' +
      '</div>';
    }).join('') || '<div style="color:var(--success);text-align:center;padding:20px;">✓ No serious defaulters</div>';
  }
}

function exportReportsCSV() {
  if (!_reportsData) { alert('Load reports first'); return; }
  var rows = [['Course','Students','Collected (₹)']];
  (_reportsData.courseWise || []).forEach(function(c) {
    rows.push([c.course, c.studentCount, c.collected]);
  });
  rows.push([]);
  rows.push(['Serious Defaulters']);
  rows.push(['Roll No','Name','Course','Months Overdue','Due (₹)','Contact']);
  (_reportsData.seriousDefaulters || []).forEach(function(s) {
    rows.push([s.id, s.name, s.course, s.monthsOverdue, s.due, s.contact]);
  });
  var csv = rows.map(function(r){ return r.join(','); }).join('\n');
  var a = document.createElement('a');
  a.href = 'data:text/csv;charset=utf-8,' + encodeURIComponent(csv);
  a.download = 'edunova-report-' + (_reportsData.year) + '.csv';
  document.body.appendChild(a); a.click();
  setTimeout(function(){ document.body.removeChild(a); }, 100);
}

File: backend/routes/reports.py
from fastapi import APIRouter, Depends, Query
from datetime import date, datetime
from collections import defaultdict
from middleware.auth_middleware import require_auth
from services.payment_service import get_all_payments_v2
from services.student_service import get_all_students

router = APIRouter(prefix="/api/reports", tags=["reports"])

MONTHS = ["Jan","Feb","Mar","Apr","May","Jun","Jul","Aug","Sep","Oct","Nov","Dec"]


@router.get("")
def get_reports(year: int = Query(default=None), user=Depends(require_auth)):
    if not year:
        year = date.today().year

    students = get_all_students()
    payments = get_all_payments_v2()
    paid = [p for p in payments if p.get("status") == "Paid"]

    today = date.today()
    this_m = f"{today.year}-{str(today.month).zfill(2)}"
    if today.month > 1:
        last_m = f"{today.year}-{str(today.month - 1).zfill(2)}"
    else:
        last_m = f"{today.year - 1}-12"

    this_month     = sum(int(p.get("amount", 0)) for p in paid if p.get("date", "").startswith(this_m))
    last_month     = sum(int(p.get("amount", 0)) for p in paid if p.get("date", "").startswith(last_m))
    total_collected = sum(int(p.get("amount", 0)) for p in paid)
    total_due       = sum(int(s.get("due", 0)) for s in students)

    # Monthly admissions for selected year
    monthly_counts = [0] * 12
    for s in students:
        created = s.get("createdAt") or s.get("enrollDate") or s.get("admissionDate") or ""
        if created and str(year) in str(created):
            try:
                m = int(str(created)[5:7]) - 1
                if 0 <= m < 12:
                    monthly_counts[m] += 1
            except Exception:
                pass

    monthly_admissions = [{"month": m, "count": c} for m, c in zip(MONTHS, monthly_counts)]

    # Course-wise breakdown
    course_stats = defaultdict(lambda: {"studentCount": 0, "collected": 0})
    for s in students:
        c = s.get("course", "")
        if c:
            course_stats[c]["studentCount"] += 1
    for p in paid:
        c = p.get("course", "")
        if c:
            course_stats[c]["collected"] += int(p.get("amount", 0))

    course_wise = sorted(
        [{"course": k, **v} for k, v in course_stats.items()],
        key=lambda x: x["studentCount"], reverse=True
    )

    # Serious defaulters — due > 0 and enrolled 2+ months ago
    serious = []
    for s in students:
        due = int(s.get("due", 0))
        if due <= 0:
            continue
        created = s.get("createdAt") or s.get("enrollDate") or ""
        months_overdue = 0
        if created:
            try:
                enroll = datetime.fromisoformat(str(created)[:10]).date()
                months_overdue = max(0, (today - enroll).days // 30)
            except Exception:
                pass
        if months_overdue >= 2:
            serious.append({
                "id":           s.get("id", ""),
                "name":         s.get("name", ""),
                "course":       s.get("course", ""),
                "due":          due,
                "monthsOverdue": months_overdue,
                "contact":      s.get("contact") or s.get("fatherMobile") or s.get("phone") or "",
            })

    serious.sort(key=lambda x: x["monthsOverdue"], reverse=True)

    return {
        "year":               year,
        "thisMonth":          this_month,
        "lastMonth":          last_month,
        "totalCollected":     total_collected,
        "totalDue":           total_due,
        "monthlyAdmissions":  monthly_admissions,
        "courseWise":         course_wise,
        "seriousDefaulters":  serious[:20],
    }
